### PR TITLE
[2.7] bpo-35757: subprocess.Popen: optimize close_fds for Linux

### PR DESCRIPTION
**Pull request by kolyshkin**
In case `close_fds=True` is passed to `subprocess.Popen()`
or its users (`subprocess.call()` etc), it might spend
some considerable time closing non-opened file descriptors,
as demonstrated by the following snippet from `strace`:

	close(3)                    = -1 EBADF (Bad file descriptor)
	close(5)                    = -1 EBADF (Bad file descriptor)
	close(6)                    = -1 EBADF (Bad file descriptor)
	close(7)                    = -1 EBADF (Bad file descriptor)
	...
	close(1021)                 = -1 EBADF (Bad file descriptor)
	close(1022)                 = -1 EBADF (Bad file descriptor)
	close(1023)                 = -1 EBADF (Bad file descriptor)

This happens because the code in `_close_fds()` iterates from 3 up to
`MAX_FDS = os.sysconf("SC_OPEN_MAX")`.

Now, syscalls are cheap, but `SC_OPEN_MAX` (also known as `RLIMIT_NOFILE`
or `ulimit -n`) can be quite high, for example:

	$ docker run --rm python python3 -c \
		$'import os\nprint(os.sysconf("SC_OPEN_MAX"))'
	1048576

This means a million syscalls before spawning a child process, which
can result in major delay, like 0.1s as measured on my fast and mostly
idling laptop. Here is the comparison with python3 (which do not have
this problem):

	$ docker run --rm python python3 -c $'import subprocess\nimport time\ns = time.time()\nsubprocess.check_call([\'/bin/true\'], close_fds=True)\nprint(time.time() - s)\n'
	0.0009245872497558594

	$ docker run --rm python python2 -c $'import subprocess\nimport time\ns = time.time()\nsubprocess.check_call([\'/bin/true\'], close_fds=True)\nprint(time.time() - s)\n'
	0.0964419841766

The solution (for Linux, at least) is easy: instead of iterating through
all the _possibly_ opened files, get the list of _actually_ opened
files (from `/proc/self/fd/`) and iterate through these. In case it
doesn't work (for example when /proc not available or some draconian
selinux/apparmor restrictions), fall back to older method.

[v2: do not use bare exceptions]

https://bugs.python.org/issue35757

### CODE REVIEW

File: Lib/subprocess.py [hunk 1]
```diff
@@ -869,6 +869,27 @@ def pipe_cloexec(self):
...
 
         def _close_fds(self, but):
+            if sys.platform.startswith("linux"):
+                # Try closing the actual opened fds (obtained via /proc),
+                # as this is way faster than iterating through MAX_FD
+                try:
+                    for f in os.listdir("/proc/self/fd"):
+                        try:
+                            fd = int(f)
+                        except:
```
Comment (asvetlov) [inline Lib/subprocess.py]: Please be precise on the expected exception type.
`except TypeError` here maybe?

File: Lib/subprocess.py [hunk 2]
```diff
@@ -869,6 +869,27 @@ def pipe_cloexec(self):
...
+                            continue
+                        if fd < 3 or fd == but:
+                            continue
+
+                        try:
+                            os.close(fd)
+                        except:
```
Comment (asvetlov) [inline Lib/subprocess.py]: Avoid bare except

File: Lib/subprocess.py [hunk 3]
```diff
@@ -869,6 +869,27 @@ def pipe_cloexec(self):
...
+                            pass
+                    return
+                except:
```
Comment (asvetlov) [inline Lib/subprocess.py]: Bare except

**Review by asvetlov — COMMENTED** [review]
Please avoid bare excepts.
I know old code uses it but there is no reason to follow this practice in new code.

File: Lib/subprocess.py (hunk 1 above)
Comment (kolyshkin) [reply Lib/subprocess.py]: Mostly `int()` returns `ValueError`, but sometimes also `TypeError`, so I guess both.

File: Lib/subprocess.py (hunk 2 above)
Comment (kolyshkin) [reply Lib/subprocess.py]: fixed

File: Lib/subprocess.py (hunk 3 above)
Comment (kolyshkin) [reply Lib/subprocess.py]: fixed

**Comment by kolyshkin** [comment]
@asvetlov Thanks for your review (as this is my first code in Python that is longer than two lines your review is very much appreciated)!  I believe I have addressed your comments. Ideally, I'd write code without using exceptions at all if something like `close_noex()` would exist, but I was not able to find anything like this.

**Comment by asvetlov** [comment]
Oooh, missed the fact that the patch is for Python 2.7

@benjaminp would you accept it or the change doesn't worth potential problems?
Python 3 has completely different implementation for closing open FDs, for the note.

**Comment by kolyshkin** [comment]
> Python 3 has completely different implementation for closing open FDs, for the note.

...which is very similar to this implementation (i.e. it also iterates through entries found in `/proc/self/fd`), except its written in C so no exceptions to care about :)

**Comment by benjaminp** [comment]
Thank you for the PR. Closing per comment on the issue.

**Pull request closed without merging**